### ISSUE TITLE
Discuss: similarities with Hypothes.is

### ISSUE THREAD

**Issue by ankostis** [karlicoss/promnesia#201]
While fiddling with tags(#195 & #199), i'm reminded with the similarities across the 2 projects: promnesia & hypothes.is [client](https://github.com/hypothesis/client) & [API](https://github.com/hypothesis/h)  (hint: i discovered promnesia through hypothes.is).

Would it make sense to consider merge promnesia extension into hypothes.is client?
I'm describing the rough features missing from each project in order to achieve that:

- hypoths.is client: 
  - decorating URLs
  - reading annotations from multiple web-annotation API backends
- hypoths.is API:
  -   respond to annotations for non-canonical URLs
- promnesia server:
  - the web-annotations API (serve only) 
  - (alternatively) implement also the write part of the the web-annotations API, so it's not necessary to extend H.client to support multiple API backends (but i'd love to see that h.client feature)

That way, promnesia UX would benefit from the man-power of hypothesis-client - eventually promnesia should focus on harvesting user-generated content, and serve it through the web-annotation API.

(i hope i'm not insulting anyone, it's just that i wish to preserve open-source developer's efforts)

**Comment by karlicoss** [comment]
> i hope i'm not insulting anyone, it's just that i wish to preserve open-source developer's efforts

Absolutely not insulting! I even have a [special label for this](https://github.com/karlicoss/promnesia/issues?q=is%3Aopen+is%3Aissue+label%3Acan-we-share%3F), and I definitely would like to collaborate with projects like Hypothesis or Worldbrian Memex. My perfect scenario is that I don't have to do anything myself at all :) But seems that we're still far from it.

In terms of what can be reused:

- In general: I basically think of backend (HPI + URL extractors in promnesia) and frontend (extension) as almost separate projects
  So I'd love to be able to swap out any of them to get something useful.

- URL normalization algorithm should 100% be shared with other projects. 
  There are many projects that could potentially benefit from it (I have more [on this page](https://beepb00p.xyz/exobrain/projects/cannon.html) ) 
  I guess it kind of addresses "respond to annotations for non-canonical URLs" suggestion? But either way I think even if Hypothes.is chooses not to support this, it should be possible to have something that works on top, e.g. syncing list of annotated pages in Hypothesis and computing canonicals on the client side.

- API: I'm not sure if there is much point of reusing it
  The Promnesia HTTP API is pretty minimal, and honestly, only exists becase you can't query an sqlite database directly from javascript. 
  On the other hand it might makes some sense to query Hypothesis API from the promnesia client (e.g. so you can use it without the backend). But it would be trickier to get right, e.g. would need some caching mechanism, extra thinking about the sources system, authorization, etc., etc.
  In this sense, perhaps Worldbrain Memex (or its Storex project) is perhaps a better candidate. But need to think more about it.

- 'decorating' -- I guess by this you mean actual highlight (i.e. matching text against the page). 
  Yep. I chatted to Oliver from Worldbrain Memex and apparently it's possible to use Hypothes.is algorithm for highlighting as a library. Frankly the only reason I haven't done it straightaway is because the current algorithm is pretty primitive and I wanted to implement it quickly to get something working as opposed to spending time integrating (you know with third party libraries there are always some, especially when it's JS ecosystem). So I kept it primitive so far specifically with the intent to integrate with Hypothes.is or something like that :)

- UX/UI: 100% agree that would be nice to benefit from some collaboration. 
  Although I have some opinions on making it highly configurable (via CSS/JS, etc.), so would like not to sacrifice that. But suggestions/experiments are always welcome. Or maybe we figure out a way to use separate/multiple UIs, would be even better.

**Comment by ankostis** [comment]
>     * 'decorating' -- I guess by this you mean actual highlight (i.e. matching text against the page).

No, i mean the rectangular boxes around links - that's something hyptohes.is client does not do at all - it only highlight annotations on trhe current page, not outgoing links.

**Comment by karlicoss** [comment]
Ah right. Yeah, it's a custom Promnesia feature (and I haven't seen anything similar anywhere else). But definitely would be cool if Hypothesis wants to support it, I want to see it in more apps :)

**Comment by ankostis** [comment]
Mentioned this discussion in [zulip "zotero" topic](https://memex.zulipchat.com/#narrow/stream/279598-general/topic/zotero/near/229270624).

**Comment by karlicoss** [comment]
Cleaned up and updated the page on Cannon, in particular about projects we could share with: https://beepb00p.xyz/exobrain/projects/cannon.html#prjctsthtcldbnftfrmt
